api/src/auth/services/auth.service.ts

```diff
@@ -59,6 +32,28 @@ export class AuthService {
         return this.generateJwt(rest);
     }
 
+    async login(userCredentialsDto: UserCredentialsDto) {
+        const { id } = await this.userService.findOneByUsername(
+            userCredentialsDto.username
+        );
+
+        return this.generateJwt({ ...userCredentialsDto, id });
+    }
+
+    async validateUser(userCredentialsDto: UserCredentialsDto) {
+        const { password } = await this.userService
+            .selectPassword(userCredentialsDto.username)
+
+        const isPasswordValid = await this.comparePasswords(
+            userCredentialsDto.password,
+            password
+        );
+
+        if (!isPasswordValid) throw new UnauthorizedException();
```

Comment:
Because this is my function which wrap bcrypt, i handle exception only inside, because if there anything go wrong, that's cause inside something gone wrong, i catch it inside for handle it immediately, thats simply function for compare hashed password, if anything go wrong that's issue with bcrypt